integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+            break;
+        case TMNOFLAGS:
+            remappingFunction = this::start;
+            break;
+        case TMRESUME:
+            remappingFunction = LocalXAResource::resume;
+            break;
+        default:
+            // Bad flags.
+            throw (XAException) new XAException(XAER_INVAL)
+                .initCause(new IllegalArgumentException("xid: " + xid + "; flags: " + flagsToString(flags)));
+        }
+
+        this.computeAssociation(XARoutine.START, xid, remappingFunction);
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "start");
+        }
+    }
+
+    // (Remapping BiFunction, used in start() above and supplied to computeAssociation() below.)
+    private Association start(Xid x, Association a) {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "start", new Object[] {x, a});
+        }
+        assert x != null; // x has already been vetted and is known to be non-null
+        if (a != null) {
+            throw new UncheckedXAException((XAException) new XAException(XAER_DUPID)
+                                           .initCause(new IllegalArgumentException("xid: " + x + "; association: " + a)));
+        }
+        Connection c;
+        try {
+            c = this.connectionFunction.apply(x);
+        } catch (RuntimeException e) {
+            // Weirdly, XAER_RMERR seems to be the "please retry" error code in this one case, not XAER_RMFAIL:
+            // https://github.com/jbosstm/narayana/blob/8ccaf0f85c7a76c227941d26cc3aa3fa9f05b160/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L682-L689.
+            throw new UncheckedXAException((XAException) new XAException(XAER_RMERR).initCause(e));
+        }
+        if (c == null) {
+            // Weirdly, XAER_RMERR seems to be the "please retry" error code in this one case, not XAER_RMFAIL:
+            // https://github.com/jbosstm/narayana/blob/8ccaf0f85c7a76c227941d26cc3aa3fa9f05b160/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L682-L689.
+            throw new UncheckedXAException((XAException) new XAException(XAER_RMERR)
+                                           .initCause(new NullPointerException("connectionFunction.apply(" + x + ")")));
+        }
+        a = new Association(Association.BranchState.ACTIVE, x, c);
+        if (LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.logp(Level.FINE, this.getClass().getName(), "start",
+                        "Created new Association ({0}) for connection ({1}) in state ACTIVE", new Object[] {a, c});
+        }
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "start", a);
+        }
+        return a;
+    }
+
+    @Override // XAResource
+    public void end(Xid xid, int flags) throws XAException {
```

Comment:
When the transaction manager decides that whatever has gone on on the current thread should stop being associated with the global JTA transaction, it will call this method.  Once again the flags tell you what sort of operation is really being invoked: a disassociation-from-the-current-thread due to an ordinary transaction boundary, a problem or a suspension.